File: src/client/controller/subpages/SettingsController.java
package client.controller.subpages;

import client.controller.ClientApplicationController;
import client.model.subpages.AvatarSelectionModel;
import client.model.subpages.SettingsModel;
import client.view.subpages.AvatarSelectionView;
import client.view.ClientApplicationView;
import client.view.prompts.ChangePassErrorView;
import client.view.prompts.ChangeProfInfoErrorView;
import client.view.prompts.PassChangeSuccessView;
import client.view.prompts.ProfChangeSuccessView;
import client.view.subpages.HomeView;
import client.view.subpages.SettingsView;
import server.model.DataPB;
import shared.SwingResources;

import javax.swing.*;
import java.awt.event.*;
import java.sql.SQLException;

public class SettingsController {
    private SettingsView settingsView; // the settings view
    private SettingsModel model; // the settings model
    private ClientApplicationView parentView; // the client application view
    private ClientApplicationController parentController; // the client application controller
    private ProfChangeSuccessView profChangeSuccessView; // the profile changed success prompt view
    private ChangeProfInfoErrorView changeProfInfoErrorView; // the profile changed failed prompt view
    private PassChangeSuccessView passChangeSuccessView; // the password changed success prompt view
    private ChangePassErrorView changePassErrorView; // the password changed failed prompt view
    private AvatarSelectionView avatarSelectionView; // the avatar selection view
    private HomeView homeView; // the home view

    public SettingsController(HomeView homeView, SettingsView view, SettingsModel model, ClientApplicationView parentView, ClientApplicationController parentController) throws SQLException {
        this.homeView = homeView;
        this.settingsView = view;
        this.model = model;
        this.parentView = parentView;
        this.parentController = parentController;

        // set texts
        this.settingsView.setFullNameText(model.getFullName());
        this.settingsView.setAvatarImagePath(model.getPFPOfUser(model.getUsername()));
        updateGameStats();

        // action listeners
        this.settingsView.setChangeAvatarListener(new ChangeAvatarListener());
        this.settingsView.setEditListener(new EditFullNameListener());
        this.settingsView.setSaveChangesListener(new SaveChangesListener());
        this.settingsView.setChangePassListener(new ChangePassListener());
        this.settingsView.setMusicListener(new MusicButtonListener());

        // mouse listeners
        this.settingsView.getBtnChangeAvatar().addMouseListener(new SwingResources.CursorChanger(view.getBtnChangeAvatar()));
        this.settingsView.getBtnEditChanges().addMouseListener(new SwingResources.CursorChanger(view.getBtnEditChanges()));
        this.settingsView.getBtnSaveChanges().addMouseListener(new SwingResources.CursorChanger(view.getBtnSaveChanges()));
        this.settingsView.getBtnChangePass().addMouseListener(new SwingResources.CursorChanger(view.getBtnChangePass()));
        this.settingsView.getBtnDelAcc().addMouseListener(new SwingResources.CursorChanger(view.getBtnDelAcc()));
        this.settingsView.getBtnMusic().addMouseListener(new SwingResources.CursorChanger(view.getBtnMusic()));

        // focus listeners

        this.settingsView.revalidate();
        this.settingsView.repaint();
        this.parentView.revalidate();
        this.parentView.repaint();
    }

    public void updateGameStats() {
        this.settingsView.setGamesPlayedText(model.getMatchesPartTwo());
        this.settingsView.setGamesWonText(model.getWinsPartTwo());
        this.settingsView.setTotalPointsText(model.getUserPoints());
    }

    class ChangeAvatarListener implements ActionListener {
        @Override
        public void actionPerformed(ActionEvent e) {
            AvatarSelectionView view = new AvatarSelectionView(homeView, settingsView);
            new AvatarSelectionController(view,
                    new AvatarSelectionModel(model.getUsername(), model.getWfImpl()));
            avatarSelectionView = new AvatarSelectionView(homeView , settingsView);
        }
    }

    class EditFullNameListener implements ActionListener {
        @Override
        public void actionPerformed(ActionEvent e) {
            settingsView.getFullNameTextField().setEnabled(true);
            settingsView.getFullNameTextField().setText("");

            Timer timer = new Timer(10000, new ActionListener() {
                @Override
                public void actionPerformed(ActionEvent e) {
                    settingsView.getFullNameTextField().setText(model.getFullName());
                    settingsView.getFullNameTextField().setEnabled(false);
                }
            });
            timer.setRepeats(false);
            timer.start();
        }
    }

    class SaveChangesListener implements ActionListener {
        @Override
        public void actionPerformed(ActionEvent e) {
            String newFullName = settingsView.getFullNameTextField().getText();
            boolean success = false;

            try {
                success = model.editInfo(model.getUsername(), "fullName", newFullName);
            } catch (SQLException ex) {
                throw new RuntimeException(ex);
            }

            if (success) {
                settingsView.setFullNameText(newFullName);
                settingsView.getFullNameTextField().setEnabled(false);
                profChangeSuccessView.main();
            } else {
                changeProfInfoErrorView.main();
            }
        }
    }

    class ChangePassListener implements ActionListener {
        @Override
        public void actionPerformed(ActionEvent e) {
            String newPassword = settingsView.getNewPassword();
            String confirmPassword = settingsView.getConfirmPassword();

            if (newPassword.equals(confirmPassword)) {
                try {
                    boolean success = model.editPassword(model.getUsername(), settingsView.getCurrentPassword(), newPassword);
                    if (success) {
                        settingsView.clearPasswordFields();
                        passChangeSuccessView.main();
                    } else {
                        settingsView.clearPasswordFields();
                        changePassErrorView.main();
                    }
                } catch (Exception ex) {
                    ex.printStackTrace();
                }
            } else {
                settingsView.getErrorMessageLabel().setVisible(true);
                settingsView.clearPasswordFields();
                changePassErrorView.main();

                Timer timer = new Timer(4000, new ActionListener() {
                    @Override
                    public void actionPerformed(ActionEvent e) {
                        settingsView.getErrorMessageLabel().setVisible(false);
                    }
                });
                timer.setRepeats(false);
                timer.start();
            }
        }
    }

    class MusicButtonListener implements ActionListener {
        @Override
        public void actionPerformed(ActionEvent e) {
            String currentState = settingsView.getMusicState();

            if (currentState.equals("ON")) {
                settingsView.setMusicState("OFF");
                parentController.stopMusic();
            } else {
                settingsView.setMusicState("ON");
                parentController.playDefaultMusic();
            }
        }
    }
}

File: src/Client_Python/yuen_client/Controller/subpages/SettingsController.py
'''
from client.controller import ClientApplicationController
from client.model.subpages import SettingsModel
from client.view.subpages import AvatarSelectionView, SettingsView
from client.view import ClientApplicationView
from client.view.promtps import ChangePassErrorView, ChangeProfInfoErrorView, PassChangeSuccessView, ProfChangeSuccessView

form server.model import DataPB
form shared import SwingResources

class SettingsController:
    def __init__(self, view, model, parentView, parentModel):
        self.view = view
        self.model = model

    def change_avatar(self):
        self.avatar_selection_view = AvatarSelectionView(self.model.get_username(),dataPB)

    def edit_full_name(self):
        self.view.full_name_entry.config(state='normal')
        self.view.full_name_entry.delete(0, tk.END)
        self.parent_view.after(10000, lambda: self.reset_full_name_entry())

    def save_changes_later(self):
        new_full_name = self.view.full_name_entry.get()
        success = self.model.edit_info(self.model.get_username(), "fullName", new_full_name)

        if success:
            print("FullName Change Success!")
            self.view.full_name_text.set(new_full_name)
            self.view.full_name_entry.config(state='disabled')
            self.prof_change_success_view = ProfChangeSuccessView()
        else:
            print("FullName Change Failed!")
            self.change_prof_info_error_view = ChangeProfInfoErrorView()

    def change_pass(self):
        new_password = self.view.new_password_entry.get()
        confirm_password = self.view.confirm_password_entry.get()

        if new_password == confirm_password:
            success = self.model.edit_password(self.model.get_username(), self.view.current_password_entry.get(), new_password)
            if success:
                print("Change Password Success!")
                self.view.clear_password_fields()
                self.pass_change_success_view = PassChangeSuccessView()
            else:
                print("Change Password Failed!")
                self.view.clear_password_fields()
                self.change_pass_error_view = ChangePassErrorView()
        else:
            self.view.error_label.config(text="Passwords do not match!")
            self.view.clear_password_fields()
            print("Passwords Do Not Match!")
            self.change_pass_error_view = ChangePassErrorView()
            self.parent_view.after(4000, self.view.hide_error_label)

    def toggle_music(self):
        current_state = self.view.music_state.get()
        if current_state == "ON":
            self.view.music_state.set("OFF")
            self.parent_controller.stop_music()
        else:
            self.view.music_state.set("ON")
            self.parent_controller.play_default_music()
'''